rest/blip_client_test.go

```diff
@@ -145,60 +152,184 @@ func NewBlipTesterClient(bt *BlipTester) (client *BlipTesterClient, err error) {
+	}
+
+	r.initHandlers(btc)
+
+	return r, nil
+}
+
+// NewBlipTesterClient returns a client which emulates the behaviour of a CBL client over BLIP.
+func NewBlipTesterClient(rt *RestTester) (client *BlipTesterClient, err error) {
+	btc := BlipTesterClient{
+		rt:   rt,
+		docs: make(map[string]map[string][]byte),
+	}
+
+	id, err := uuid.NewRandom()
+	if err != nil {
+		return nil, err
 	}
 
-	bt.blipContext.DefaultHandler = func(request *blip.Message) {
-		btc.storeMessage(request)
-		panic("unexpected request to DefaultHandler")
+	if btc.pushReplication, err = newBlipTesterReplication("push"+id.String(), &btc); err != nil {
+		return nil, err
+	}
+	if btc.pullReplication, err = newBlipTesterReplication("pull"+id.String(), &btc); err != nil {
+		return nil, err
 	}
 
 	return &btc, nil
 }
 
-// Start will begin a continous replication since 0 between the client and server
-func (btc *BlipTesterClient) Start() {
-	btc.StartSince("true", "0")
+// StartPull will begin a continuous pull replication since 0 between the client and server
+func (btc *BlipTesterClient) StartPull() (err error) {
+	return btc.StartPullSince("true", "0")
 }
 
-// Start will begin a replication between the client and server with the given params.
-func (btc *BlipTesterClient) StartSince(continous, since string) {
+// StartPullSince will begin a pull replication between the client and server with the given params.
+func (btc *BlipTesterClient) StartPullSince(continuous, since string) (err error) {
 	getCheckpointRequest := blip.NewRequest()
 	getCheckpointRequest.SetProfile(messageGetCheckpoint)
-	getCheckpointRequest.Properties[blipClient] = btc.id
-	if !btc.bt.sender.Send(getCheckpointRequest) {
-		panic("unable to send getCheckpointRequest")
+	getCheckpointRequest.Properties[blipClient] = btc.pullReplication.id
+	if err := btc.pullReplication.sendMsg(getCheckpointRequest); err != nil {
+		return err
 	}
 
 	subChangesRequest := blip.NewRequest()
 	subChangesRequest.SetProfile(messageSubChanges)
-	subChangesRequest.Properties[subChangesContinuous] = continous
+	subChangesRequest.Properties[subChangesContinuous] = continuous
 	subChangesRequest.Properties[subChangesSince] = since
 	subChangesRequest.SetNoReply(true)
-	if !btc.bt.sender.Send(subChangesRequest) {
-		panic("unable to send subChangesRequest")
+	if err := btc.pullReplication.sendMsg(subChangesRequest); err != nil {
+		return err
 	}
+
+	return nil
 }
 
-// Close will close the underlying BlipTester, and remove the stored docs and messages.
+// Close will empty GetServerUUIDthe stored docs, close the underlying replications, and finally close the rest tester.
 func (btc *BlipTesterClient) Close() {
-	btc.bt.Close()
 	btc.docsLock.Lock()
 	btc.docs = make(map[string]map[string][]byte, 0)
 	btc.docsLock.Unlock()
 
-	btc.messagesLock.Lock()
-	btc.messages = make(map[blip.MessageNumber]*blip.Message, 0)
-	btc.messagesLock.Unlock()
+	btc.pullReplication.Close()
+	btc.pushReplication.Close()
+
+	btc.rt.Close()
```

Comment:
As discussed, don't have BTC close the rest tester - leave that to the test itself.